### DISCUSSION TITLE
Question about non-zero BEC values for originally missing entries in `bec_train.out`, and the meaning of “single-phase” in qNEP training

### DISCUSSION THREAD

**Discussion by zx25j** [brucefan1983/GPUMD#1421] [Q&A]
Hello, I am using the NEP training workflow in GPUMD to train a qNAP force field, and I encountered a couple of questions that I hope anyone could clarify.

### 1. Non-zero BEC values for originally missing BEC entries

During training, I noticed that in the output file `bec_train.out`, some atoms whose BEC values were originally missing in my dataset ended up with non-zero BEC components in the output.
After checking more carefully, I noticed a clear pattern:
**these unexpected non-zero BEC values always appear on the last three atoms of each frame.**

For example:

``` shell
-2.1982 0.0330041 -0.0230443 -0.00528663 -4.5384 0.072697 -0.0200977 0.0979562 -2.16074 0 0 0 0 0 0 0 0 0
-2.16532 0.0741293 -0.0143186 0.0437055 -4.72404 0.0267158 -0.0134018 0.0650209 -2.15282 0 0 0 0 0 0 0 0 0
-2.21249 0.103773 -0.0339843 0.0570327 -4.72548 0.00854517 -0.0361527 0.0387659 -2.18393 0 0 0 0 0 0 0 0 0
-2.14201 0.0333167 -0.00981977 -0.0043423 -4.64656 0.00315027 -0.0085634 0.0554635 -2.13306 0 0 0 0 0 0 0 0 0
-2.18555 0.010423 2.02724e-05 -0.0232763 -4.54478 0.0261997 -0.00271125 0.0732561 -2.15377 0 0 0 0 0 0 0 0 0
-2.1982 0.0861354 -0.0108376 0.0402559 -4.66829 0.0206886 -0.0145276 0.051137 -2.17175 0 0 1.52601e-42 0 4.15478 4.04497 3.927 8.20191 8.27817
-2.18166 0.0950673 -0.042636 0.0589512 -4.7694 0.0896439 -0.0375761 0.108433 -2.18459 8.18604 12.2242 12.2768 0.02276 4.11503 4.02675 4.25748 8.07157 8.18192
-2.15443 0.0764058 -0.0172804 0.042443 -4.54273 0.0145216 -0.018684 0.0600936 -2.1593 8.37163 12.2461 12.2957 0.10704 4.1054 4.18819 4.16505 8.21057 8.1774
1.26352 0.0333994 0.0122098 0.0309348 1.26197 0.00375088 -0.00730828 0.00479309 1.25794 0 0 0 0 0 0 0 0 0
1.23331 0.0193475 0.018949 0.0171536 1.25254 0.00964879 0.0199836 0.0084432 1.23123 0 0 0 0 0 0 0 0 0
1.25942 0.0171515 0.0215993 0.0156838 1.25524 -0.00466854 0.0199248 -0.00894314 1.23744 0 0 0 0 0 0 0 0 0
1.2641 0.0146771 0.0131707 0.00563403 1.26741 0.0145335 0.0118406 0.00782662 1.25403 0 0 0 0 0 0 0 0 0
1.232 0.0219099 0.00958384 0.0220932 1.23621 0.0162137 0.0126267 0.0118321 1.23306 0 0 0 0 0 0 0 0 0
```

This made me wonder whether this is:

* an expected behavior of the training/output procedure,
* caused by some issue in my input settings or data preparation,
* or possibly an unintended behavior/bug.

### 2. Meaning of “single-phase”

According to the manual, qNEP currently only applies to **single-phase**. I would like to better understand what exactly “single-phase” means in this context.

Does “single-phase” refer to:

* one crystallographic phase only,
* no coexistence of different phases,
* or a structurally homogeneous dataset with the same parent phase?

In the qNAP-related literature, for example in the BTO case, would the **ROTC** phases still be considered a single-phase ?

### Additional context

My main concern is whether my dataset/setup violates the intended usage conditions for qNEP, and whether that could explain the unexpected non-zero BEC values in `bec_train.out`.

I would appreciate any clarification or guidance. Thank you very much.

**Comment by brucefan1983** [comment]
1) if you have BEC data for the atoms 6-8, their values are not correct I think. For example the reference values 
```
0 0 1.52601e-42 0 4.15478 4.04497 3.927 8.20191 8.27817
```
does not look right to me.

2) Columns 1-9 are predicted ones and 10-18 are reference values. The code will always predict, even for structures without BEC references.

3) singel phase material means a system with a well defined $\epsilon_{\infty}$.

**Reply by zx25j** [reply]
Thank you very much for your reply, Professor Fan.

I apologize for not explaining my previous example clearly enough. The first eight lines actually correspond to the **last eight atoms of one structure**, while the following lines already belong to the **atomic properties of another structure**.

In addition, I re-checked the relevant structures in the training set and can confirm that these structures themselves **do not have BEC properties**. However, I found that a structure originally containing **132 atoms** is written as having **135 atomic entries** in `bec_train.out`. Moreover, the non-zero BEC entries consistently appear on the **last three atoms of each frame**.

This might be related to my **batch-size setting**. As described in the manual, when I use the **prediction mode**, this phenomenon disappears. However, I am still confused about whether these non-zero values appearing here could affect the calculation of the **loss function**. I am also unsure whether the extra **three lines** are related to the fact that the maximum number of atoms in my training set is **135**.

In addition, I have uploaded the `train.xyz` file together with the corresponding **two-frame files** from `bec_train.out` as attachments for reference. Thank you again for your help.

[train.xyz.txt](https://github.com/user-attachments/files/26310034/train.xyz.txt)
[bec_train.out.txt](https://github.com/user-attachments/files/26310038/bec_train.out.txt)

**Comment by brucefan1983** [comment]
I think you were using a non-fullbatch training, in which case, the output order in bec_train.out is not the same as in trian.xyz. You can make a prediction by running with `prediction 1` to get the order right.
